### PR TITLE
Extends virtual module astro:transitions/client to export swapFunctions

### PR DESCRIPTION
**Pull request by martrapp**
## Changes

This is a follow up to #10908.

The creation of the blocks for user-defined swap functions has proven to be very useful, e.g. for https://github.com/stackblitz/tutorialkit. 

This PR makes it possible to import the^se functions via `astro:transitions/client`. 

## Testing

Adapted the existing e2e tests to use the virtual module

## Docs

/cc @withastro/maintainers-docs for feedback!

The `swapFunctions` object exported here provides access to functions that you can use to define your own `swap()` function as a replacement in View Transitions. These functions have not yet been documented in the docs. This is a very specific functionality that is probably only relevant for very few users. 

If we decide to explicitly document them in the docs, I could certainly use some help.

### CODE REVIEW

**Review by matthewp — APPROVED** [review]

**Comment by delucis** [comment]
> These functions have not yet been documented in the docs. This is a very specific functionality that is probably only relevant for very few users.

Even if it’s an advanced API, we should still document it! Would you be able to kick things off with a draft docs PR and we discuss it there @martrapp?

**Comment by martrapp** [comment]
> Even if it’s an advanced API, we should still document it! Would you be able to kick things off with a draft docs PR and we discuss it there @martrapp?

... examining AD^2 to make the docs PR ...

**Comment by martrapp** [comment]
> Even if it’s an advanced API, we should still document it! Would you be able to kick things off with a draft docs PR and we discuss it there @martrapp?

Opened https://github.com/withastro/docs/pull/9084 for discussion.

File: .changeset/new-monkeys-sit.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': patch
```
Comment (ematipico) [inline .changeset/new-monkeys-sit.md:2]:
```suggestion
'astro': minor
```

**Review by ematipico — DISMISSED** [review]

**Comment by ematipico** [comment]
Sorry for the block, but since we are exposing a **new API** to userland, this counts as a minor because, for the users, it's a new functionality. In fact, it requires docs :)

**Comment by martrapp** [comment]
Hi Ema, no problem, I really should have figured that out myself!
At least this makes it easy to guess the right version for `<Since>` in docs 😄

**Review by ematipico — APPROVED** [review]

File: .changeset/new-monkeys-sit.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': minor
+---
+
+Extends `astro:transitions/client` with `swapFunctions`, which provides building blocks for custom swap functions.
```
Comment (sarah11918) [inline .changeset/new-monkeys-sit.md:5]:
``````suggestion
Adds a new object `swapFunctions` to expose the necessary utility functions on `astro:transitions/client` that allow you to build custom swap functions as view transitions.

The example below uses these functions to recreate Astro's own default `swap` transition as a defined custom swap:

```astro
<script>
import { swapFunctions } from 'astro:transitions/client';
// substitutes window.document with doc
function myAstroSwap(doc: Document) {
  swapFunctions.deselectScripts(doc);
  swapFunctions.swapRootAttributes(doc);
  swapFunctions.swapHeadElements(doc);
  const restoreFocusFunction = swapFunctions.saveFocus();
  swapFunctions.swapBodyElement(doc.body, document.body);
  restoreFocusFunction();
};

  event.swap = () => myAstroSwap(event.newDocument);

<script>
```
``````
Some of my prepositions may be a little off here, but what would you think about something of this style? Can you make this accurate?

**Review by sarah11918 — COMMENTED** [review]
Hey @martrapp ! For a minor, you always get a beefier changeset to introduce and hype a new feature! Often, we'll pull closely from the docs for people who prefer to "get their news" from the changelog.

See what you think about something like below!

**Comment by martrapp** [comment]
Thanks for the suggestion, @sarah11918! I've made a few small changes to demonstrate something more interesting than just showing how to rebuild what's already there. Should we also include a link to the “Docs” section with this example?

File: .changeset/new-monkeys-sit.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  const oldMain = document.querySelector('main');
+  if (newMain && oldMain) {
+    swapFunctions.swapBodyElement(newMain, oldMain);
+  } else {
+    swapFunctions.swapBodyElement(doc.body, document.body);
+  }
+  restoreFocusFunction();
+};
+<script>
+```
```
Comment (sarah11918) [inline .changeset/new-monkeys-sit.md:30]:
``````suggestion
```

See the [view transitions guide](https://docs.astro.build/en/guides/view-transitions/#astrobefore-swap) for more information about hooking into the `astro:before-swap` lifecycle event and adding a custom swap implementation.
``````
What do you think about a link like this, that also is a friendly reminder of where this is, both in docs, and with respect to the lifecycle events?

File: .changeset/new-monkeys-sit.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+---
+'astro': minor
+---
+
+Adds a new object `swapFunctions` to expose the necessary utility functions on `astro:transitions/client` that allow you to build custom swap functions to be used with view transitions. 
+
+The example below uses these functions to replace the default `swap` function with one that only swaps the `<main>` part of the page:
```
Comment (sarah11918) [inline .changeset/new-monkeys-sit.md:7]:
```suggestion
The example below uses these functions to replace Astro's built-in default `swap` function with one that only swaps the `<main>` part of the page:
```

**Review by sarah11918 — APPROVED** [review]
One tiny super nit edit, and approving for docs!  :partying_face:

**Review by yanthomasdev — APPROVED** [review]

**Pull request merged**